src/live_restore/live_restore_fs.c

```diff
@@ -298,55 +300,76 @@ __live_restore_fs_directory_list_worker(WT_FILE_SYSTEM *fs, WT_SESSION *wt_sessi
-      ENOENT, false);
 
-    for (namep = dirlist_src; namep != NULL && *namep != NULL; namep++) {
-        /* Create file path for the file from the source directory. */
-        WT_ERR(__live_restore_create_file_path(session, &lr_fs->source, *namep, &temp_path));
-        WT_ERR_NOTFOUND_OK(
-          __live_restore_fs_has_file(lr_fs, &lr_fs->destination, session, temp_path, &dest_exist),
-          false);
-        WT_ERR(__dest_has_tombstone(lr_fs, temp_path, session, &have_tombstone));
-
-        if (!dest_exist && !have_tombstone) {
-            WT_ERR(__wt_realloc_def(session, &dirallocsz, count_dest + count_src + 1, &entries));
-            WT_ERR(__wt_strdup(session, *namep, &entries[count_dest + count_src]));
-            ++count_src;
-        }
+    WT_ERR(lr_fs->os_file_system->fs_exist(
+      lr_fs->os_file_system, wt_session, path_src, &source_folder_exists));
 
-        __wt_free(session, temp_path);
-        if (single)
-            goto done;
+    if (source_folder_exists) {
+        WT_ERR(lr_fs->os_file_system->fs_directory_list(
+          lr_fs->os_file_system, wt_session, path_src, prefix, &dirlist_src, &num_src_files));
+
+        for (namep = dirlist_src; namep != NULL && *namep != NULL; namep++) {
+            /*
+             * We're iterating files in the source, but we want to check if they exist in the
+             * destination, so create the file path to the backing destination file.
+             */
+            WT_ERR(
+              __live_restore_create_file_path(session, &lr_fs->destination, *namep, &temp_path));
+            WT_ERR_NOTFOUND_OK(__live_restore_fs_has_file(
+                                 lr_fs, &lr_fs->destination, session, temp_path, &dest_exist),
+              false);
+            WT_ERR(__dest_has_tombstone(lr_fs, temp_path, session, &have_tombstone));
+
+            if (!dest_exist && !have_tombstone) {
+                WT_ERR(
+                  __wt_realloc_def(session, &dirallocsz, count_dest + count_src + 1, &entries));
+                WT_ERR(__wt_strdup(session, *namep, &entries[count_dest + count_src]));
+                ++count_src;
+            }
+
+            __wt_free(session, temp_path);
+            if (single)
+                goto done;
+        }
     }
 
+    if (!dest_folder_exists && !source_folder_exists)
+        WT_ERR_MSG(session, ENOENT,
+          "Cannot report contents of '%s'. Folder does not exist in the source or destination.",
+          directory);
+
 done:
 err:
     __wt_free(session, path_dest);
     __wt_free(session, path_src);
     __wt_free(session, temp_path);
     if (dirlist_dest != NULL)
-        WT_TRET(__live_restore_fs_directory_list_free(fs, wt_session, dirlist_dest, count_dest));
+        WT_TRET(
+          __live_restore_fs_directory_list_free(fs, wt_session, dirlist_dest, num_dest_files));
     if (dirlist_src != NULL)
-        WT_TRET(__live_restore_fs_directory_list_free(fs, wt_session, dirlist_src, count_src));
```

Comment:
`count_src` only tracks files present in the source and not present or tombstoned in the destination. This meant on cleanup we didn't free the whole directory list for the source directory and we upset ASan.